### PR TITLE
WIP: State manager middleware

### PR DESCRIPTION
**Pull request by michalczaplinski**
#### Description of what you did:

WIP on adding middleware to `@frontity/connect`

- [ ] add context to actions
- [ ] add user warning when detecting a breaking reference
- [ ] add the middleware (this should be broken down more)

#### My PR is a:

- 🔝 Improvement

#### Is the PR ready to be merged?

- 🚧 Work in progress

### CODE REVIEW

**Comment by luisherranz** [comment]
I haven't had time to look into it in detail but looks good to me so far!

For async middleware, my idea is to allow async functions, but the modification of the mutation, the modification of the action arguments and the abortion must be synchronous (before any `await`).

Something like this: https://codesandbox.io/s/naughty-ives-v5kzn 

Let me know if this is useful or not.

File: packages/connect/src/__tests__/connect.tests.jsx [hunk 1]
```diff
@@ -154,4 +154,26 @@ describe("connect", () => {
     );
     expect(app).toMatchSnapshot();
   });
+
```
Comment (michalczaplinski) [inline packages/connect/src/__tests__/connect.tests.jsx:157]: First test for the middleware :)

File: packages/connect/src/connect.js [hunk 2]
```diff
@@ -34,6 +34,7 @@ export function connect(Comp) {
   const isStatelessComp = !(Comp.prototype && Comp.prototype.isReactComponent);
 
   let ReactiveComp;
+  const componentName = Comp.displayName || Comp.name;
```
Comment (michalczaplinski) [inline packages/connect/src/connect.js:37]: we'll attach it to the context in the middleware

File: packages/connect/src/connect.js [hunk 3]
```diff
@@ -45,7 +46,10 @@ export function connect(Comp) {
       const [, setState] = useState();
 
       // get frontity from the context;
-      const frontity = useContext(context);
```
Comment (michalczaplinski) [inline packages/connect/src/connect.js:48]: just renamed: `s/frontity/rawStore`

File: packages/connect/src/create-store.js [hunk 4]
```diff
@@ -19,36 +21,49 @@ const getSnapshot = obj => {
...
+      newObj[key] = convertedActions(obj[key], instance, context);
       return newObj;
     }, {});
   }
 };
 
-export const createStore = store => {
-  const observableState = observable(store.state);
+export const createStore = (store, context) => {
+  // TODO: onl throw in dev ?
```
Comment (michalczaplinski) [inline packages/connect/src/create-store.js:53]: We can later add a `__DEV__` mode to `connect` to make use of those warnings. React does something similar, they have many warnings like: 
```js
if (__DEV__) { throw new Error(`Terrible error`) } 
```
and use a webpack (or rollup, not sure) plugin to remove those if statements completely for production.

File: packages/connect/src/internals.js [hunk 5]
```diff
@@ -1,3 +1,23 @@
 export const proxyToRaw = new WeakMap();
-export const rawToProxy = new WeakMap();
+// export const rawToProxy = new WeakMap();
 export const rawToRoot = new WeakMap();
```
Comment (michalczaplinski) [inline packages/connect/src/internals.js]: `rawToRoot` should be removed. We can store a global reference to the root somewhere.

**Review by michalczaplinski — COMMENTED** [review]
@luisherranz Here are some updates to the PR. You're very welcome to have a look and comment (just keep in mind it's  still WIP)

**Comment by luisherranz** [comment]
Closing this for now. We'll review them later once we are back with a final API for the connection middleware.

**Pull request closed without merging**